src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/builder/BigQuerySQLBuilder.java

```diff
@@ -106,28 +189,127 @@ protected void appendFieldEqualityClause(JoinDefinition joinDefinition) {
     appendFieldEqualityClause(stages, stageNameToJoinKeyMap, joinOnNullKeys);
   }
 
+  @VisibleForTesting
   protected void appendFieldEqualityClause(List<JoinStage> stages,
                                            Map<String, JoinKey> stageNameToJoinKeyMap,
                                            boolean joinOnNullKeys) {
     JoinStage prev = null;
 
     for (JoinStage curr : stages) {
       if (prev == null) {
-        String currStage = curr.getStageName();
-        builder.append(getFullTableName(currStage)).append(" AS ").append(getTableAlias(currStage));
+        appendFullTableNameAndAlias(curr.getStageName());
       } else {
-        builder.append(" ");
+        builder.append(SPACE);
         appendJoinOnKeyOperation(prev, curr, stageNameToJoinKeyMap, joinOnNullKeys);
       }
 
       prev = curr;
     }
   }
 
+  private void appendOnExpressionClause() {
+    JoinCondition.OnExpression onExpression = (JoinCondition.OnExpression) joinDefinition.getCondition();
+
+    JoinStage left = joinDefinition.getStages().get(0);
+    JoinStage right = joinDefinition.getStages().get(1);
+
+    // Append Join Statement for these 2 stages
+    // ...<left_table> <join_type> JOIN <left_table> ON ...
+    appendFullTableNameAndAlias(left.getStageName());
+    builder.append(SPACE);
+    appendJoinStatement(left, right);
+
+    String leftAlias = stageToTableAliasMap.get(left.getStageName());
+    String rightAlias = stageToTableAliasMap.get(right.getStageName());
+
+    String leftStageIdentifier = onExpression.getDatasetAliases()
+      .getOrDefault(left.getStageName(), left.getStageName());
+    String rightStageIdentifier = onExpression.getDatasetAliases()
+      .getOrDefault(right.getStageName(), right.getStageName());
+
+    appendOnExpressionClause(onExpression.getExpression(),
+                             leftStageIdentifier, leftAlias,
+                             rightStageIdentifier, rightAlias);
+  }
+
+  /**
+   * Adds expression clause for these 2 stages.
+   * <p>
+   * Executes replace operation to replace table names in the supplied string for BigQuery-safe table aliases.
+   *
+   * @param expression       expression to manipulate
+   * @param leftIdentifier   Left Stage name
+   * @param leftReplacement  Left Stage replacement alias
+   * @param rightIdentifier  Right Stage Name
+   * @param rightReplacement Right Stage replacement alias
+   */
+  @VisibleForTesting
+  protected void appendOnExpressionClause(String expression,
```

Comment:
this is also the behavior in the Spark implementation, it just uses the user given input as-is.

Reply:
My idea with the re-write was to ensure no conditions can cause a syntax error in BQ (as Spark and BQ have slightly different syntax rules).

I agree that, for simplicity, it would be easier to just paste the input as-is, but I wouldn't like for the user to run into a condition where preview works but then BQ fails.

Reply:
But, to your point, SparkSQL will be as picky as BQ is will something like a non-ascii dataset/table name..

Reply:
After some further testing, I realized your point is pretty valid. Since I'm already checking that stage names and stage aliases do not contain the escape characters for spark ( ` ) and BigQuery ( \ ), supplying the raw SQL without modifications should not cause any problems in BigQuery that were not present in Spark.